File: app/src/main/java/com/geekorum/ttrss/data/SynchronizationDao.java
/**
 * Geekttrss is a RSS feed reader application on the Android Platform.
 *
 * Copyright (C) 2017-2018 by Frederic-Charles Barthelery.
 *
 * This file is part of Geekttrss.
 *
 * Geekttrss is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Geekttrss is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Geekttrss.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.geekorum.ttrss.data;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

/**
 * Dao to access the database when synchronizing articles
 */
@Dao
public abstract class SynchronizationDao {
    @Query("SELECT * FROM transactions")
    public abstract List<Transaction> getAllTransactions();

    @Delete
    public abstract void deleteTransaction(Transaction transaction);

    @Query("SELECT * FROM articles WHERE _id=:id")
    public abstract Article getArticleById(long id);

    @Query("SELECT _id FROM articles ORDER BY _id DESC LIMIT 1")
    public abstract long getLatestArticleId();

    @Update
    public abstract void updateArticle(Article article);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertCategories(List<Category> categories);

    @Query("DELETE FROM ARTICLES where feed_id=:feedId")
    abstract void deleteArticleFromFeed(long feedId);

    @Query("SELECT * FROM categories")
    public abstract List<Category> getAllCategories();

    @Delete
    public abstract void deleteCategories(List<Category> categories);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertFeeds(List<Feed> feeds);

    @Query("SELECT * FROM feeds")
    public abstract List<Feed> getAllFeeds();

    @Delete
    abstract void deleteFeeds(List<Feed> toBeDelete);

    @androidx.room.Transaction
    public void deleteFeedsAndArticles(List<Feed> toBeDelete) {
        for (Feed feed : toBeDelete) {
            deleteArticleFromFeed(feed.getId());
        }
        deleteFeeds(toBeDelete);
    }

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertArticles(List<Article> dataArticles);

    @Query("UPDATE articles SET unread=:unread, transiant_unread=:transientUnread, marked=:starred, published=:published,"
            + " last_time_update=:lastTimeUpdated, is_updated=:isUpdated"
            + " WHERE _id=:id")
    public abstract void updateArticleMetadata(long id, boolean unread, boolean transientUnread, boolean starred,
                                        boolean published, long lastTimeUpdated, boolean isUpdated);
}

File: app/src/main/java/com/geekorum/ttrss/data/SynchronizationDao.kt
/*
 * Geekttrss is a RSS feed reader application on the Android Platform.
 *
 * Copyright (C) 2017-2022 by Frederic-Charles Barthelery.
 *
 * This file is part of Geekttrss.
 *
 * Geekttrss is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Geekttrss is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Geekttrss.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.geekorum.ttrss.data

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update

/**
 * Dao to access the database when synchronizing articles
 */
@Dao
abstract class SynchronizationDao {
    @Query("SELECT * FROM transactions")
    abstract suspend fun getAllTransactions(): List<Transaction>

    @Query("SELECT _id FROM articles ORDER BY _id DESC LIMIT 1")
    abstract suspend fun getLatestArticleId(): Long?

    @Query("SELECT _id FROM articles WHERE feed_id=:feedId ORDER BY _id DESC LIMIT 1")
    abstract suspend fun getLatestArticleIdFromFeed(feedId: Long): Long?

    @Query("SELECT * FROM categories")
    abstract suspend fun getAllCategories(): List<Category>

    @Query("SELECT * FROM feeds")
    abstract suspend fun getAllFeeds(): List<Feed>

    @Query("SELECT * FROM feed_fav_icon")
    abstract fun getAllFeedFavIcons(): List<FeedFavIcon>

    @Delete
    abstract suspend fun deleteTransaction(transaction: Transaction)

    @Query("SELECT * FROM articles WHERE _id=:id")
    abstract suspend fun getArticleById(id: Long): Article?

    @Query("SELECT * FROM articles WHERE feed_id=:feedId ORDER BY feed_id DESC LIMIT 1")
    abstract fun getLatestArticleFromFeed(feedId: Long): Article?

    @Update
    abstract suspend fun updateArticle(article: Article)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract suspend fun insertCategories(categories: List<Category>)

    @Query("DELETE FROM ARTICLES where feed_id=:feedId")
    internal abstract suspend fun deleteArticleFromFeed(feedId: Long)

    @Delete
    abstract suspend fun deleteCategories(categories: List<Category>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract suspend fun insertFeeds(feeds: List<Feed>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract suspend fun setFeedFavIcon(feedFavIcon: FeedFavIcon)

    @Delete
    internal abstract suspend fun deleteFeeds(toBeDelete: List<Feed>)

    @Query("DELETE FROM feed_fav_icon where _id=:feedId")
    internal abstract suspend fun deleteFeedFavIcon(feedId: Long)

    @androidx.room.Transaction
    open suspend fun deleteFeedsAndRelatedData(toBeDelete: List<Feed>) {
        for ((id) in toBeDelete) {
            deleteArticleFromFeed(id)
            deleteFeedFavIcon(id)
        }
        deleteFeeds(toBeDelete)
    }

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract suspend fun insertArticles(dataArticles: List<Article>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract suspend fun insertArticlesTags(articlesTags: List<ArticlesTags>)

    @Update(entity = Article::class)
    abstract suspend fun updateArticlesMetadata(metadata: List<Metadata>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract suspend fun insertAttachments(attachments: List<Attachment>)
}
